pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            // Process the orchestrator chain block author (if it exists) and activity tracking is enabled
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                total_weight.saturating_accrue(T::DbWeight::get().reads(1));
+                if let ActivityTrackingStatus::Enabled { start, end: _ } =
+                    <CurrentActivityTrackingStatus<T>>::get()
+                {
+                    total_weight.saturating_accrue(T::DbWeight::get().reads(1));
+                    if start <= T::CurrentSessionIndex::session_index() {
+                        total_weight
+                            .saturating_accrue(Self::on_author_noted(orchestrator_chain_author));
+                    }
+                }
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn process_ended_session() {
+            let current_session_index = T::CurrentSessionIndex::session_index();
+            // Since this function will be executed in the beginning of a session
+            // we can populate the active collators from the session that just ended
+            // before resetting the collators storage for the current session
+            // without affecting the current session's active collators records
+            ActiveCollators::<T>::insert(
+                current_session_index.saturating_sub(1),
+                <ActiveCollatorsForCurrentSession<T>>::get(),
+            );
+
+            <ActiveCollatorsForCurrentSession<T>>::put(BoundedBTreeSet::new());
+
+            // Cleanup active collator info for sessions that are older than the maximum allowed
+            if current_session_index > T::MaxInactiveSessions::get() {
+                <crate::pallet::ActiveCollators<T>>::remove(
+                    current_session_index
+                        .saturating_sub(T::MaxInactiveSessions::get())
+                        .saturating_less_one(),
+                );
+            }
+        }
+        pub fn on_author_noted(author: T::CollatorId) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+            let _ = <ActiveCollatorsForCurrentSession<T>>::try_mutate(
+                |active_collators| -> DispatchResult {
+                    if active_collators
+                        .try_insert(author)
+                        .map_err(|_| Error::<T>::MaxCollatorsPerSessionReached)?
+                    {
+                        total_weight.saturating_accrue(T::DbWeight::get().writes(1));
+                    }
+                    Ok(())
+                },
+            );
+            total_weight
+        }
+    }
+}
+
+impl<T: Config> NodeActivityTrackingHelper<T::CollatorId> for Pallet<T> {
+    fn is_node_inactive(node: &T::CollatorId) -> bool {
```

Comment:
Depending on how this function is called it may return invalid results. For example, if a new collator joins, it has not produced any blocks in the past sessions, so this function will say it is inactive.

So I guess the only way for this to work is to keep track of which collators have been collators for more than 5 sessions, and only allow checking activity for those. Am I right or am I missing something?

Reply:
No there is something we can do, which is count non-activity instead of activity. that way you can clearly distinguish between a collator not joining and one being inactive

Reply:
@chexware revisit tomasz suggestion, and if it's true, the let's chain the code to count whether a collator was `offline` for a given session (I think for that you just need to change the `process_ended_session` function)

Reply:
@chexware Agree that should prevent marking newly  joined collators as inactive.

Reply:
Using inactivity tracking from https://github.com/moondance-labs/tanssi/pull/922/commits/dc58e94a1504539385be07e11cb57d075cef1955 . However as a result the integration tests are braking

Reply:
Tests related issues fixed in https://github.com/moondance-labs/tanssi/pull/922/commits/22d672ccf9180a94d92e11cadf8a9965675ce75e and https://github.com/moondance-labs/tanssi/pull/922/commits/7c91fc87fd8f6c64b508d37ee5bca2e61fbad624